test/math_libraries/convolutions.py

```diff
@@ -0,0 +1,113 @@
+                bias_g = self._subtensor(bias, 0, group, g)
+                outputs.append(self._thnn_conv(input_g, weight_g, k, bias_g, s, p, d))
+            return torch.cat(outputs, 1)
+
+    def _collect_cases(self):
+        dir_name = 'convolutions_cases'
+        cur_dir = os.getcwd()
+        search_command = 'ls {}/{}/*'.format(cur_dir, dir_name)
+        case_files = os.popen(search_command).read().strip().split()
+        total_cases = list()
+        for file_name in case_files:
+            with open(file_name) as f:
+                model_cases = json.loads(f.read())
+            total_cases += model_cases
+        return total_cases
+
+    # Set multiple situations to improve convolution2d test coverage
+    @onlyOnCPUAndCUDA
+    @skipCUDAIfRocm
+    @dtypesIfCUDA(torch.half, torch.float, torch.double)
+    @dtypes(torch.float)
+    def test_conv2d_ext(self, device, dtype):
+        # this list for save the cases which dh,dw == 0
+        # and raise error in the end
+        invalid_cases = list()
+
+        total_cases = self._collect_cases()
+        for case in total_cases:
+            case_name = case['case_name']
+            bs = case['mb']
+            group = case['g']
+            ic, ih, iw = case['ic'], case['ih'], case['iw']
+            oc = case['oc']
+            kh, kw = case['kh'], case['kw']
+            sh, sw = case['sh'], case['sw']
+            ph, pw = case['ph'], case['pw']
+            dh, dw = case['dh'], case['dw']
+            has_bias = case['bias']
+            if dh == 0 or dw == 0:
+                invalid_cases.append(case_name)
+                continue
+
+            ic_g = ic // group
+            torch.manual_seed(1)
+            input = torch.randn((bs, ic, ih, iw), device=device, dtype=dtype, requires_grad=True)
+            weight = torch.randn((oc, ic_g, kh, kw), device=device, dtype=dtype) * 0.01
+            bias = None if has_bias == 'False' else torch.randn((oc), device=device, dtype=dtype)
+
+            k = [kh, kw]
+            s = [sh, sw]
+            p = [ph, pw]
+            d = [dh, dw]
+            thnn_output = self._thnn_conv_group(input, weight, k, bias, s, p, d, group)
+            if self.device_type == 'cpu' and torch.backends.mkldnn.is_available():
+                output = torch.mkldnn_convolution(input, weight, bias, p, s, d, group)
+            elif self.device_type == 'cuda' and torch.backends.cudnn.is_available():
+                output = torch.cudnn_convolution(input, weight, bias, p, s, d, group, True, True)
+            else:
+                output = torch.conv2d(input, weight, bias, s, p, d, group)
+
+            msg = 'device:{}, dtype:{}, group:{}, batchsize:{}' \
+                  'input channel:{}, output channel:{}, ' \
+                  'bias:{}, padding:{}, dilation:{}, stride:{}, ' \
+                  'kernel:{}'
+            msg = msg.format(device, dtype, group, bs, ic, oc, has_bias, p, d, s, k)
+
+            if self.device_type == 'cpu':
```

Comment:
This conditional block has two identical clauses. Maybe you can remove the if/elif?